### DISCUSSION TITLE
Does sum_expr work with vectors?

### DISCUSSION THREAD

**Discussion by alexcharn5** [dirkschumacher/rmpk#74] [Q&A]
I'm trying to set a model objective, and I have a matrix A with the same number of rows as the number of variables, which are binary. And basically I want to take a columnwise average using only the rows that end up being chosen (x = 1). In this example 3 rows will be chosen. My attempt (shown below) at doing this involves using sum_expr on rows of A; the _m_ th row would only contribute to the sum if x[_m_] = 1. But I get an error. Do you have a suggested way of handling this?

k=3
x <- model$add_variable("x", type="binary", i=1:n_models)
model$set_objective(sum_expr(mean(k/sum_expr(A[m,]*x[m],m=1:n_models)*A[i,])*x[i],i=1:n_models))

**Comment by dirkschumacher** [comment]
Quick answer: it is not vectorized. You could do something like
`sum_expr(A[i, j] * x[k], i = 1:n, j = 1:m, k = 1:k)`. You also cannot use the function `mean`. Hope this helps a bit.

**Reply by alexcharn5** [reply]
Hmm, I've been able to use `mean` inside of sum_expr before.

Alternatively, is there a way to "get" the value of the variable inside sum_expr, i.e., treat it like a normal variable? For example:
`sum_expr(mean(apply(sweep(A,1,x,'*'),2,sum)*B[i,])*x[i],i=1:n)`

**Reply by dirkschumacher** [reply]
`mean` might work in certain cases as it uses the operators defined for `rmpk` variables I believe, but it might not do what you expect. `rmpk` in its current form is designed to support functions in [summation form](https://en.wikipedia.org/wiki/Summation). I would suggest to define your functions in terms of basic operators `+`, `-`, `*`, `/` and `^`. `sum_expr` is meant to construct a sum over a set of indexes.
